### PR TITLE
plumbing: resolve non-external delta references

### PR DESCRIPTION
**Pull request by ZauberNerd**
In a self-contained pack file delta references might point to base
objects stored later in the file.
In this case we need to replace placeholders for external refs with the
actual base object and update the children references.

Fixes: #484

### CODE REVIEW

**Comment by mcuadros** [comment]
Can you add a test for this?

**Comment by ZauberNerd** [comment]
I have prepared a test fixture here: https://github.com/go-git/go-git-fixtures/pull/6  I guess that needs to be merged first before I can write a test here.

**Comment by ericzhill** [comment]
I just ran into this problem today and can confirm this PR does fix the issue for me as well.

**Comment by abrander** [comment]
I too can confirm that this PR solves an issue with current master.

Exhibit:

```go
git.PlainClone("/tmp/repo", true, &git.CloneOptions{
	URL: "https://github.com/influxdata/line-protocol",
	NoCheckout: true,
})
```

**Comment by kyparisisg** [comment]
I am running to the same problem and it looks like this PR solves the it!
could you please consider merging it?

**Comment by qliklars** [comment]
We have also run into this problem. Is there a reason for not merging the PR?

**Comment by pjbgf** [comment]
It would be great if there were tests that could reproduce the error this fixes.

**Comment by per1234** [comment]
@pjbgf isn't there one here?: https://github.com/go-git/go-git-fixtures/pull/6

**Comment by ZauberNerd** [comment]
@pjbgf yes, that is how I think I'll do the tests. But for that to happen, the fixture needs to be merged first: https://github.com/go-git/go-git/pull/485#issuecomment-1114061954
Otherwise please advise me how to proceed or do the change yourself. Maintainer edits are allowed on this PR.

**Comment by pjbgf** [comment]
@ZauberNerd I had a look at this today and can confirm that using your test fixture (and one additional test), this behaviour is fixed and tested for regression.

Please add another test in `parser_test.go`, which is similar to `TestResolveExternalRefsInThinPack`, but refers to the new tag `delta-before-bash` you have created on https://github.com/go-git/go-git-fixtures/pull/6. 

Whilst the other PR is not merged you can add this to the `go.mod`:
```
// Temporary fix, waiting for https://github.com/go-git/go-git-fixtures/pull/6.
replace github.com/go-git/go-git-fixtures/v4 => github.com/ZauberNerd/go-git-fixtures/v4 v4.0.0-20220316085525-96d6741f0b44
```

That would allow for the tests to pass, and for a maintainer to quickly test the entire E2E of this PR.

Once the other PR is merged, the above will need to be removed and the `go-git-fixtures` dependency updated.

**Comment by oshoval** [comment]
Hi, Thanks for this great PR
Is it possible please to advance it ?
We also have the problem without it with some of our repositories, which blocks parts of the automation
Thanks

**Comment by qliklars** [comment]
We also have problems without it, so it would be great if it was merged.
Thanks!

**Comment by Deiz** [comment]
Not to pile on too much, but is there anything blocking merge of this and https://github.com/go-git/go-git-fixtures/pull/6 which it depends on, @mcuadros @pjbgf?

Anecdotally, something appears to have shifted on GitLab's side but non-external delta references seem to be cropping up with vastly higher frequency in GitLab-hosted repositories. Our organization has hit them twice in the last week, and it'd be nice to not litter a bunch of replace directives around to take advantage of this fix. 🙂

**Comment by pjbgf** [comment]
I can't speak on what is blocking the merge, as the changes looks good to me. However, I don't have write access on this repository nor any official saying on the matter.

I did merge these changes on a fork I use (back in December) and so far it seems to be working as intended. So you are probably better off forking the project, and cherry-picking it until it gets merged.

**Comment by becjon** [comment]
> Not to pile on too much, but is there anything blocking merge of this and [go-git/go-git-fixtures#6](https://github.com/go-git/go-git-fixtures/pull/6) which it depends on, @mcuadros @pjbgf?
> 
> Anecdotally, something appears to have shifted on GitLab's side but non-external delta references seem to be cropping up with vastly higher frequency in GitLab-hosted repositories. Our organization has hit them twice in the last week, and it'd be nice to not litter a bunch of replace directives around to take advantage of this fix. 🙂

We also have the problems you describe in combination with GitLab.com

File: go.mod
Comment (pjbgf) [inline go.mod:33]: The other PR has been merged, please remove the `replace` and update the `go-git-fixtures` in the `require` section to:

```
github.com/go-git/go-git-fixtures/v4 v4.3.2-0.20230305113008-0c11038e723f
```

**Review by pjbgf — CHANGES_REQUESTED** [review]
@ZauberNerd can you please rebase and check the nits below.

**Comment by pjbgf** [comment]
> I don't have write access on this repository nor any official saying on the matter.

This has now changed, so I should be able to get this merged once it is rebased.

**Comment by oshoval** [comment]
Gm, any chance this can be advanced please ?
Thanks

**Comment by ZauberNerd** [comment]
@mcuadros @pjbgf sorry for the delay. I have rebased and updated the PR. Please take another look.

**Comment by pjbgf** [comment]
The [current failure](https://github.com/go-git/go-git/actions/runs/4504167118/jobs/8211140251?pr=485) is not related to the change. I will fix that on a follow-up PR.

**Review by pjbgf — APPROVED** [review]
@ZauberNerd Thank you very much for working on this and following up with the rebase.

**Comment by oshoval** [comment]
Thanks, this is awesome
Is a new release planned please ?

**Comment by pjbgf** [comment]
@oshoval in the beginning of next month we should have another version cut.

**Comment by oshoval** [comment]
Hi, it seems no new release yet ? can you please consider one ?
Thanks

**Comment by oshoval** [comment]
@pjbgf 
Thanks 

FYI 
https://sonarcloud.io/project/security_hotspots?id=kubevirt_cluster-network-addons-operator&pullRequest=1561&resolved=false&types=SECURITY_HOTSPOT

**Pull request merged**
